llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

```diff
@@ -11019,8 +11039,6 @@ BoUpSLP::getEntryCost(const TreeEntry *E, ArrayRef<Value *> VectorizedVals,
 
   bool NeedToShuffleReuses = !E->ReuseShuffleIndices.empty();
   if (E->isGather()) {
-    if (allConstant(VL))
-      return 0;
```

Comment:
This code should remain as is, update your instructions target costs accordingly

Reply:
Without removing this check, it wont go into `getBuildVectorCost`

Reply:
So, with the change i.e. after removal of this check, it goes into BasicTTI for X86/RISCV which returns 0 as it previously did. Is there anything incorrect here?

Reply:
It should return 0. TTI handles it in getArithemticInstrCost/etc. by checking operands. If the operand is constant and can be folded away, it is free, otherwise these function include the cost of the constant materialization, where required

Reply:
you mean I should call the API `getConstMaterializationCost` here?

Reply:
No, you should call it in AARCH64TargetTransformInfo where required

Reply:
I am not able to follow you. Could you tell how the above check should look like and from where I should call `getConstMaterializationCost`?

Reply:
Check RISCVTargetTransformInfo.cpp, RISCVTTIImpl::getArithmeticInstrCost function. It has lambda `getConstantMatCost` which is called if Op1Info.isConstant() is true or Op2Info.isConstant() is true to add the constants materialization costs, if operands are constants

Reply:
1. To say, `getConstantMatCost` in RISCV is not doing detailed analysis about the constants involved. Its very primitive in nature. For AArch64, I need to know about the constants involved.
2. If you are worried about tests failing for X86/RISCV/other targets by removing this check, this is not happening because I am returning 0 from some other place(i.e BasicTTI).

And in short, nothing is affected for other targets.

Reply:
1. I do not see any kind of "complexity" in your constants analysis. OperandValueInfo already contains most of the stuff required. If you need some extra stuff, expand OperandValueInfo/Operands info.
2. Adding a call here introduces an extra divergence point, which complicates things.

Reply:
> I do not see any kind of "complexity" in your constants analysis.

I intend to do that in the next revision of the patch.


I would like to note down the advantages/disadvantages of having constant materialization API consistent with x86/RISCV  i.e. just pass the operands to the TTI API who will call the  constMaterialization API

Advantages:
1. Consistency with other targets
2. Minimal changes in SLP

Disadvantages in the context of this patch:
1. There are multiple scenarios in SLP where constant vectors are used e.g. arithmetic instructions, memory instructions like store, etc. Passing operands for each of the scenario is redundancy and tedious and maybe this needs modification to the TTI APIs as well(getMemoryOpCost for e.g.)
2. When operands are passed to the  TTI APIs, I need to call `getConstantMaterializationCost` from such APIs as is case with RISCV. So, this is redundancy and tedious in the backend as well.
3. The cost for constant vector will be shown to be 0 like
   SLP: Adding cost 0 for bundle Idx: 2, n=2 [i32 11, ..].

   In actual, you would be adding the cost for this constant vector alongwith the instruction. This is confusing and hard to debug.
4. For this specific case, I need to keep track of the vector constants that are already materialized per function so that I avoid double counting the cost at places like loops. Having code for keeping track of already materialized const vectors is easy in SLP rather than keeping in TTI backend and its also the piece of code which is not cost-modelling per se.

Just passing the operands to TTI backend should entail having this piece of code in the backend.

Reply:
On the other hand, if I remove this code, I am able to calculate the const cost using a single call and a single SPI. This is lot too easy.